snuba/migrations/migrations.py

```diff
@@ -0,0 +1,91 @@
+import logging
+
+from datetime import datetime
+from enum import Enum
+from importlib import import_module
+from typing import Optional, Sequence, Tuple
+
+from snuba.clusters.cluster import ClickhouseClientSettings, get_cluster, CLUSTERS
+from snuba.clusters.storage_sets import StorageSetKey
+from snuba.migrations.operations import Operation
+
+
+logger = logging.getLogger("snuba.migrations")
+
+LOCAL_TABLE_NAME = "migrations_local"
+DIST_TABLE_NAME = "migrations_dist"
+
+# Currently only local migrations are supported
+TABLE_NAME = LOCAL_TABLE_NAME
+
+
+class App(Enum):
+    SYSTEM = "system"
+    SNUBA = "snuba"
```

Comment:
What is the goal for splitting system and snuba migrations? It seem to me we are not able to enforce dependencies between snuba and system ones.

Reply:
I will\ make this clearer in the implementation but the idea is that each migration defines a list of dependencies in the format (app, migration_id). So the first snuba migration will have a dependency on our system migration ("system", "0001_migrations") . The goal for splitting them into groups was to provide a way to define certain groups of migrations as optional in future, e.g. those associated with a particular optional feature could be skipped. However the dependencies will still be run if a migration in another group that is active depends on it.

Reply:
It seems we would be introducing a lot of complexity by adding the support for optional features, since the migration engine would not really support them here (while a developer could already use dependencies). I would put this out of scope unless we have a clear design around how optional features would work. How would a migration be defined as optional? How would the user select that they want that feature as well? How would we ensure that all migrations for an optional feature are executed from the beginning of time? Would we have optional features depending on other optional features? 
Unless we are planning to answer all these questions now I would put them out of scope.

Reply:
The way I imagined this to work from the user perspective is something like this:
- There is no such concept of an optional marker on a migration or group of migrations per se
- Scenario 1: default scenario - the user can choose to run all of the migrations. We run all the migrations in all the apps ensuring that any dependencies are executed before the ones that depend on them.
- Scenario 2: the user can pass a migration ID. The runner walks the chain of dependencies and runs all the migrations that were directly or indirectly specified as dependencies there. Anything else will be skipped.
- Scenario 3: the user decides which "apps" they want to migrate, either via CLI  args or via settings. We run all the migrations in those apps and all of their dependencies. Anything else will be skipped.

Reply:
re: scenario 2: doesn't this mean (as of now) all the migrations in their sequential order up to the provided one, for all the apps ?

I think there are a few tricky corner cases if we start having independent sequences of migrations. Let's say that I am at version 9, between version 9 and 19 there is a blocking migration at version 16 for a specific app and that app is not a mandatory one.
I move up to 14 without ever migrating that app. Then I move to 19 still skipping the app that has the blocking migration on version 16.
Now I am at version 19 and I want to start using that app, so I need to apply all its migrations, these require me to stop at 16, but I am already passed that version. 
When we were designing this process we were assuming one sequence of migrations, if we drop this assumption, we need to revisit that all the scenarios we designed are still viable.